service/worker/scheduler/workflow.go

```diff
@@ -726,6 +753,38 @@ func (s *scheduler) sleep(nextWakeup time.Time) {
 	}
 }
 
+func (s *scheduler) processBackfills() {
+	limit := s.tweakables.BackfillsPerIteration
+
+	for len(s.State.OngoingBackfills) > 0 &&
+		limit > 0 &&
+		// use only half the buffer for backfills
+		len(s.State.BufferedStarts) < s.tweakables.MaxBufferSize/2 {
```

Comment:
If we have a maximum size why not use a circular buffer for `s.State.OngoingBackfills` so we don't ever need to reallocate it. Dropping the front reduces the capacity, meaning we'll need to reallocate regularly

Reply:
In general I don't expect people to do more than one backfill at once. That seems like a premature optimization. If there's some existing data structure that would let the code remain as simple but have better allocation behavior, we could switch to that.

Reply:
When do backfills happen, only when a schedule is created? I don't consider good data structure selection to be premature optimization, but if this is a one-off thing that happens then I care far less

Reply:
It happens whenever a user requests it by api call, but it is generally a one-off or few-off thing, not regular.

Anyway, I did look into replacing it with a linked list. The problem is that the slices (BufferedStarts and OngoingBackfills) are in a proto-generated struct (that's used as the current state, to make continue-as-new easy). So we'd need to shuffle them back and forth... possible, but it's getting more complicated.

Overall I think the overhead of some extra allocation is going to be in the noise considering workflow tasks, SDK overhead, etc. If it was an every-iteration thing, then it'd be worth paying more attention.

Reply:
I agree wholly. Let’s ship it